---
title: Exploratory analysis
author: Jeff Leek
output:
  rmarkdown::html_document:
    toc: true
  vignette: >  
    %\VignetteIndexEntry{The three tables in genomics}
    %\VignetteEngine{knitr::rmarkdown}
    \usepackage[utf8]{inputenc}
---

```{r front, child="./../front.Rmd", echo=FALSE}
```


## Dependencies

This document depends on the following packages:

```{r load_hidden, echo=FALSE, results="hide", warning=FALSE}
suppressPackageStartupMessages({
  library(devtools)
  library(Biobase)
  library(RSkittleBrewer)
  library(gplots)
  library(dplyr)
  library(AnnotationDbi)
})
```

```{r load}
  library(gplots)
  library(devtools)
  library(Biobase)
  library(RSkittleBrewer)
  library(org.Hs.eg.db)
  library(AnnotationDbi)
```


To install these packages you can use the code (or if you are compiling the document, remove the `eval=FALSE` from the chunk.)

```{r install_packages, eval=FALSE}
install.packages(c("devtools","gplots"))
source("http://www.bioconductor.org/biocLite.R")
biocLite(c("Biobase","org.Hs.eg.db","AnnotationDbi"))
biocLite("alyssafrazee/RSkittleBrewer")
```


### General principles

* Use plots as often as possible
* Use tables for phenotype data
* Look for: 
    * Missing values
    * Outlier values
    * Mislabeled samples
    * Naming consistency

### Make the plots pretty

Typically we will use color to explore data sets and label different values. There are a large number of color options in R. I like the [RSkittleBrewer](http://alyssafrazee.com/RSkittleBrewer.html) package, but you can also check out Jenny Bryan's excellent [lecture on colors in R](http://www.stat.ubc.ca/~jenny/STAT545A/block14_colors.html) for more information. 

Load the library and set the color palette with the `palette` function. Now when I type `col = 1` it will look for the first color in the `trop` colors. We also set the character to be a filled dot with `par(pch=19)`.

```{r pretty, eval=FALSE}
library(RSkittleBrewer)
# Make the colors pretty
trop = RSkittleBrewer("tropical")
palette(trop)
par(pch=19)
```

### Load some data

We will use this expression set to look at how we use plots and tables to check for different characteristics

```{r load_data}
con = url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bodymap_eset.RData")
load(file=con)
close(con)
bm = bodymap.eset
pdata=pData(bm)
edata=exprs(bm)
fdata = fData(bm)
ls()
```

### Tables for factor/character variables

Tables are good for looking at factor or character variables, especially in phenotype data

```{r tables}
table(pdata$gender)
table(pdata$gender,pdata$race)
```

### Look for missing values

First check a summary of the distribution to look for scale, this is also one way to check for `NA` values.

```{r summary}
summary(edata)
```

`NA` is the most common character for missing values, but sometimes they are coded as spaces, 999, -1 or "missing". Check for missing values in a variety of ways

```{r missing}
# Use option useNA to include NA's in table
table(pdata$age,useNA="ifany")

# is.na checks for NA values
table(is.na(pdata$age))

# Check for other common missing names
sum(pdata$age==" ")

# Check genomic data for NAs
sum(is.na(edata))

# Make the distribution of NA's by genes
gene_na = rowSums(is.na(edata))
table(gene_na)

# Make the distribution of NA's by samples
sample_na = rowSums(is.na(edata))
table(sample_na)

```

### Make sure dimensions match up
The number of rows of the feature data should match the number of rows of the genomic data (both are the number of genes). The number of rows of the phenotype data should match the number of columns of the genomic data (both are the number of samples).
```{r dimensions}
dim(fdata)
dim(pdata)
dim(edata)
```

### Look at overall distributions

Here we see that there are a lot of outliers
```{r boxplot}
boxplot(log2(edata+1),col=2,range=0, las=2, cex.axis=0.8)
```

We can also look at this sample by sample with histograms

```{r histograms}
par(mfrow=c(1,2))
hist(log2(edata[,1]+1),col=2, cex.main=0.8)
hist(log2(edata[,2]+1),col=2, cex.main=0.8)
```

Or with density plots

```{r densities}
plot(density(log2(edata[,1]+1)),col=2)
lines(density(log2(edata[,2]+1)),col=3)
```

A very common task is to compare distributions of measurements (say before normalization). You can do this with a qq-plot

```{r}
qqplot(log2(edata[,1]+1), log2(edata[,2]+1),col=3)
```

A very widely used plot is what is known as a [M-A plot](https://en.wikipedia.org/wiki/MA_plot), sometimes called a [Bland Altman plot](https://en.wikipedia.org/wiki/Bland%E2%80%93Altman_plot). The basic idea is to plot the sum of the two values on the x-axis and the difference on the y-axis. This can be used to see any difference between the (samples, averages, etc.) and to see if there is any intensity-specific biases. 

```{r}
mm = log2(edata[,1]+1) - log2(edata[,2]+1)
aa = log2(edata[,1]+1) + log2(edata[,2]+1)
plot(aa,mm,col=2)
```


We can remove rows that are mostly zero and notice any differences in the distributions across samples.

```{r}
edata = as.data.frame(edata)
filt_edata = filter(edata,rowMeans(edata) > 1)
boxplot(as.matrix(log2(filt_edata+1)),col=2, las=2, cex.axis=0.8)
```

### Check for obvious data mixups

Here we are going to do a check to make sure that the men and women are correctly labeled by looking at expression on the Y chromosome. In general you might do several of this type of check to confirm the data are correctly labeled.  

Get the chromosomes for each gene using the feature data. 
```{r}
aeid = as.character(fdata[,1])
chr = AnnotationDbi::select(org.Hs.eg.db,keys=aeid,keytype="ENSEMBL",columns="CHR")
head(chr)
```

Filter to the data on chromsome Y and sum up all the counts. A tricky issue is that some genes are annotated to multiple chromsomes. Here we take the first chromsome each is annotated to. 
```{r}
dim(chr)
dim(edata)
# Take non-duplicated chromsomes
chr = chr[!duplicated(chr[,1]),]

# Confirm that the annotation still is in the right order
all(chr[,1] == rownames(edata))

# Select the chromosome Y samples
edatay = dplyr::filter(edata,chr$CHR=="Y")

# Males have Y chromsome expression as expected
boxplot(colSums(edatay) ~ pdata$gender)
points(colSums(edatay) ~ jitter(as.numeric(pdata$gender)),
        col=as.numeric(pdata$gender),
        pch=19)

```


### Heatmaps

A common type of plot for genomics data is a heatmap. They are usually used for visualizing matrices. For example we can look at all genes with an average number of counts greater than 500: 

```{r}
ematrix = as.matrix(edata)[rowMeans(edata) > 10000,]
heatmap(ematrix)
```

We might change the coloring since this one is a little hard to see. To do this you have to set up a color palette. 

```{r}
colramp = colorRampPalette(c(3,"white",2))(9)
heatmap(ematrix,col=colramp)
```

You might have noticed some automatic clustering here, you can turn that off (we'll learn more about it in a later lecture)


```{r}
heatmap(ematrix,col=colramp,Rowv=NA,Colv=NA)
```

If you load the `gplots` package you can add a color scale with the `heatmap.2` package. Here we have to add some options to make the dendogram disappear, scale the data by rows, and remove a tracing plot. 

```{r}
heatmap.2(ematrix,col=colramp,Rowv=NA,Colv=NA,
          dendrogram="none", scale="row",trace="none")
```



## Session information

Here is the session information 

```{r session_info}
devtools::session_info()
```

It is also useful to compile the time the document was processed. This document was processed on: `r Sys.Date()`.

